server/src/main/java/org/apache/iotdb/db/mpp/execution/operator/process/FilterAndProjectOperator.java

```diff
@@ -0,0 +1,341 @@
+    }
+    // init output datatypes
+    outputDataTypes =
+        Arrays.stream(outputExpressions)
+            .map(expression -> typeProvider.getType(expression.getExpressionString()))
+            .collect(Collectors.toList());
+  }
+
+  // result of commonExpressions should be kept
+  private void addReferenceCountOfCommonExpressions() {
+    for (Expression expression : commonSubexpressions) {
+      if (predicateMap.containsKey(expression)) {
+        predicateMap.get(expression).addReferenceCount();
+      }
+    }
+  }
+
+  @Override
+  public OperatorContext getOperatorContext() {
+    return operatorContext;
+  }
+
+  @Override
+  public TsBlock next() {
+    TsBlock input = inputOperator.next();
+    if (input == null) {
+      return null;
+    }
+
+    // reset ColumnTransformer for next input
+    predicateColumnTransformer.reset();
+
+    if (!hasNonMappableUDF) {
+      for (ColumnTransformer columnTransformer : outputColumnTransformers) {
+        columnTransformer.reset();
+      }
+    }
+
+    TsBlock filterResult = getFilterTsBlock(input);
+
+    // contains non-mappable udf, we leave calculation for TransformOperator
+    if (hasNonMappableUDF) {
+      return filterResult;
+    }
+    return getTransformedTsBlock(filterResult);
+  }
+
+  /**
+   * Return the TsBlock that contains both initial input columns and columns of common
+   * subexpressions after filtering
+   *
+   * @param input
+   * @return
+   */
+  private TsBlock getFilterTsBlock(TsBlock input) {
+    final TimeColumn originTimeColumn = input.getTimeColumn();
+    // feed Predicate ColumnTransformer, including TimeStampColumnTransformer and constant
+    for (Expression expression : predicateMap.keySet()) {
+      if (predicateInputLocations.containsKey(expression.getExpressionString())) {
+        predicateMap
+            .get(expression)
+            .initializeColumnCache(
+                input.getColumn(
+                    predicateInputLocations
+                        .get(expression.getExpressionString())
+                        .get(0)
+                        .getValueColumnIndex()));
+      } else if (expression.getExpressionType().equals(ExpressionType.TIMESTAMP)) {
+        predicateMap.get(expression).initializeColumnCache(originTimeColumn);
+      } else if (expression.getExpressionType().equals(ExpressionType.CONSTANT)) {
+        predicateMap
+            .get(expression)
+            .initializeColumnCache(new ConstantColumn((ConstantOperand) expression));
+      }
+    }
+
+    predicateColumnTransformer.tryEvaluate();
+
+    Column filterColumn = predicateColumnTransformer.getColumn();
+
+    final TsBlockBuilder tsBlockBuilder = TsBlockBuilder.createWithOnlyTimeColumn();
+    tsBlockBuilder.buildValueColumnBuilders(predicateOutputDataTypes);
+    final TimeColumnBuilder timeBuilder = tsBlockBuilder.getTimeColumnBuilder();
+    final ColumnBuilder[] columnBuilders = tsBlockBuilder.getValueColumnBuilders();
```

Comment:
use `predicateTsBlocker.reset()` to avoid re new TsBlockBuilder each time here.
and remember to add a new reset method in `TsBlockBuilder` to pass initEntrySize here.